### ISSUE TITLE
Add option to store song list on disk

### ISSUE THREAD

**Issue by joshkunz** [joshkunz/ashuffle#16]
In order to pick a random song out of the MPD library, ashuffle stores a list of all of the user’s songs in-memory. In memory-constrained environments, users with large libraries can experience OOMs or other crashes.

I think the most straightforward implementation would just be using temporary files and something like [cdb](http://cr.yp.to/cdb.html). The source is public domain, and should be easy to embed in ashuffle.

**Comment by joshkunz** [comment]
After investigating some open-source db implementations, I think they're probably all a bit heavyweight for ashuffle's use-case. I think the most straightforward thing is to just use memory-mapped files for storing URI strings, and a simple in-memory (not memory-mapped) index. This should remove the bulk of the run-time memory cost.

**Comment by joshkunz** [comment]
For anyone looking at this issue: It's very likely that the solution will look a lot like swap. If you can enable swap on your Linux install, that will probably provide ~all the benefit right now. I still think that OOMing on large libraries when swap is not enabled is sub-par, so I'm still working on this. Just wanted to given anyone running into issues a heads up.

**Comment by martinetd** [comment]
(For low memory devices, zram currently sees a lot of attention -- playlist data is mostly text so would compress well and that might work for that usage as well.
That said file-backed mmap should be usable as memory just fine and the OS will page out part of the file as required... I think that'd be horribly slow on low end SD cards if it keeps reading from it but if you don't modify it much it should just work?)

**Comment by joshkunz** [comment]
(Why did you put your whole statement in a parenthetical? 😝 )

> For low memory devices, zram currently sees a lot of attention.

That is very neat. I had not heard of zram before. That's a cool idea. For low compute devices (like those that often have low memory), I suspect this would be slow as well. I agree that it would compress really well though, since it's mostly text. Definitely something to look into.

> That said file-backed mmap should be usable as memory just fine

This is my current plan, hence the comment above. A memory-mapped file is just an application-specific swap in this case. Agreed that SD Cards used by these single board computers are pretty slow, but we can always write it to leave the next N songs in memory so skips/enqueues are fast.

To give a little more detail on what's going on here: I'm working on quantifying the severity of the problem, and writing some memory usage tests. If we're going to guarantee some specific memory overhead, we should have testing for it. Once I can measure the memory usage of ashuffle on large libraries, I'll set a target (I'm thinking something like < 250MiB), and work on an implementation.

**Comment by joshkunz** [comment]
Alright... Well this has been an adventure. After [some missteps](http://jkz.wtf/random-linux-oddity-1-ru_maxrss), and the addition of [quite a bit of infrastructure](https://github.com/joshkunz/fakelib), I've finally been able to write an automated memory usage test (PR #77). After looking at the data, I'm pretty inclined to just say that this is actually not a problem.

In the PR, I run `ashuffle --only 1000` in 3 different scenarios:

* Library of 30k MP3s, with a minimum URI length of 71 bytes (the average URI length in my personal library).
  * Peak memory usage: `8.77 MiB`
* Library of 500k MP3s, with a minimum URI length of 71 bytes.
  * Peak memory usage: `160 MiB`
* Library of 500k MP3s, with a minimum URI length of 500 bytes.
  * Peak memory usage: `440 MiB`

Even if my guesses for typical library size and URI length are off, and these numbers are wrong by a factor of 10, the memory usage is fine. 30k tracks is a *huge* library, 500k is unheard of (for reference Spotify has ~60 Million tracks in its catalogue). The 8.77 MiB for 30k tracks with realistic path lengths is well below my 250 MiB target from the [comment above](https://github.com/joshkunz/ashuffle/issues/16#issuecomment-647097951).

My assumption in issue #14 (the issue that spawned this one), that ashuffle was getting OOM-killed was probably incorrect. Rather, #14 is probably caused by #31 which has since been fixed, and performance regression tests have been added. Unless someone has a good argument that my library size/URI lengths are too low, I'm going to close this once PR #77 is merged.

If someone one-day finds this because they're experiencing memory issues, feel free to re-open, or open a new issue. I'd be happy to revisit this if needed.

**Comment by joshkunz** [comment]
PR #77 now merged, so closing this out.
